File: strategy/generator/confirm/supertrend.py
from dataclasses import dataclass

from core.models.parameter import Parameter, StaticParameter
from core.models.source import SourceType

from .base import Confirm, ConfirmType


@dataclass(frozen=True)
class SupertrendConfirm(Confirm):
    type: Confirm = ConfirmType.Sup
    source_type: Parameter = StaticParameter(SourceType.HL2)
    atr_period: Parameter = StaticParameter(10.0)
    factor: Parameter = StaticParameter(3.0)

File: ta_lib/strategies/confirm/src/supertrend.rs
use base::prelude::*;
use core::prelude::*;
use timeseries::prelude::*;
use trend::supertrend;

pub struct SupertrendConfirm {
    source_type: SourceType,
    atr_period: usize,
    factor: f32,
}

impl SupertrendConfirm {
    pub fn new(source_type: SourceType, atr_period: f32, factor: f32) -> Self {
        Self {
            source_type,
            atr_period: atr_period as usize,
            factor,
        }
    }
}

impl Confirm for SupertrendConfirm {
    fn lookback(&self) -> usize {
        self.atr_period
    }

    fn filter(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        let (direction, _) = supertrend(
            &data.source(self.source_type),
            data.close(),
            &data.atr(self.atr_period),
            self.factor,
        );

        (direction.sgt(&ZERO), direction.slt(&ZERO))
    }
}
